### ISSUE TITLE
I want to be able to create variable fields in Lightdash

### ISSUE THREAD

**Issue by TuringLovesDeathMetal** [lightdash/lightdash#4172]
### Description

I want to be able to create dimensions which are variable in Lightdash. 

Some examples of use cases for this:

- I want a dimension to show the most profitable customer so I can use this in a KPI dashboard. 
- I want a dimension be either using the GBP revenue column or the EUR revenue column, so I can calculate the `total revenue` in either GBP or EUR depending on which filter option I pick in my dashboard
- I want to create a dimension which is a URL for each customer in hubspot (the URL is the same, but the company name would be variable depending on the company).
- I want to create a filter which is injected into the query (and runs a different underlying table query when I use a different filter value). 

### Examples of other tools

- Qlik Sense: https://share.vidyard.com/watch/bts3r8FuehLRkd2THrQ98m?
- Tableau: https://help.tableau.com/current/online/en-us/data_stories_customize_context_variables.htm (maybe this is a better link? https://interworks.com/blog/2024/02/26/tableau-hacks-dynamic-parameters/)
- Looker: https://cloud.google.com/looker/docs/liquid-variable-reference
- Metabase: https://www.metabase.com/learn/sql-questions/sql-variables#introduction-to-sql-variables-and-filter-widgets

**Comment by Kladar** [comment]
Also would love a Looker-style liquid variable reference!

**Comment by Kladar** [comment]
What do we want? "Liquid variable reference!" when do we want it? "in a reasonable timeframe!" :)

**Comment by TuringLovesDeathMetal** [comment]
Another user example: Parameterized filters would be also along the lines of a filter on a dashboard which is month vs day for example, that then switches order date to order month

**Comment by owlas** [comment]
@Kladar I loved your comment, keep the issue alive! 

There is actually some liquid available and we're adding some more options. 

@TuringLovesDeathMetal is this different from the [milestone for user attributes](https://github.com/lightdash/lightdash/milestone/92) and the existing liquid syntax on url fields for dimensions?

**Comment by Kladar** [comment]
Still relevant, we want user-defined variable definition!

**Comment by Daniel-SalazarSoplapuco** [comment]
Still Relevant

**Comment by colin-nayuta** [comment]
really want it, will improve dashboarding experience massively.
discussion here:
https://lightdash-community.slack.com/archives/C03MG2VTR08/p1712746296116979

**Comment by colin-nayuta** [comment]
imagine this use case:

Business case: Lightdash Analytics
Case 1
Metrics = number of users
Chart = Stacked Bar Chart
X Axis = Reporting date by month
Y Axis = Number of users
How to stack?
Use dynamic aggregation
Option 1: All, not agg, just one simple bar
Option 2: By Country,  UK US Russia Etc..
Option 3: By Company sector, food, saas, crypto, mobility, etc…
Option 4: By DB connectors, Snowflake, Postgres, Databrick, etc…
Option 5: By DBT versions they are using, 1.1, 1.4. 1.6777, etc…
Option 789101112….
Now you have build 5 different graphs to show the same metrics with different “stack/aggregration”
with this feature,  one graph can kill all birds.
benefit? much easier for the data partitioner for maintenance, 1 vs 5
for data consumers, 1 vs 5 too, just click and choose their favour.
Case 2: Continue case 1 + Time Dynamic aggregation, by day by month by year
same benefit, no need to build multiple graphs, just 1 kill all birds
Tableau example from above, is a hack between parameters and dimension
Looker have it too, there was a piece of Lookml code circulate everywhere, probably you can get it from some existing looker users.

**Comment by jpetey75** [comment]
Another use case for these variable fields would be for dashboard users to input a custom input, like a value multiplier, and have that input be used in a metric or table calculation. 

For example, a variable for "Profit Margin" that defaults to .3, but can be changed by a user to .5 or any other value, then refresh the dashboard to see charts updated to use the new value for various calculations. 

([Internal customer issue for reference](https://lightdash.slack.com/archives/C03MCQ08UKS/p1718092233979089))

**Comment by TuringLovesDeathMetal** [comment]
Example from customer: 

this would be useful to be able to set at dashboard level.
An example for us would be:
As an auction platform for cars, we would would like to compare one brand of car vs the rest. Where a variable will allow us to create a custom dimension where the value returned is either
- selected brand
- all other brands
the variable could then be used at dashboard or single chart level for a sales head to flick between brands to check performance

https://lightdash-community.slack.com/archives/C067S48BHRP/p1721031767410239?thread_ts=1720004640.363279&cid=C067S48BHRP

**Comment by jasonlanderson** [comment]
Yep agree, this is still relevant for our use cases too.  Thanks!

**Comment by etadelta222** [comment]
Very popular ask!

**Comment by colin-nayuta** [comment]
when will we have it? have been waiting for more than 2 years...

**Comment by aivaras-mazylis** [comment]
Really looking forward to this — would unlock a lot of powerful reporting use cases for us. Subscribed and happy to test when it's ready! 🙌

**Comment by hannaht0** [comment]
Still relevant, it would be a great feature to increase usability and value to our dashboards.

**Comment by jynl** [comment]
Is there any update on this or where does it sit on the current roadmap? Would love to see this feature!

**Comment by owlas** [comment]
It coming up soon on the roadmap and already being designed! There's a lot of use cases here, so we might not cover all of them in the first cut, but keep an eye on this issue for updates! https://www.lightdash.com/roadmap

**Comment by magnew** [comment]
https://headwayapp.co/lightdash-updates/lightdash-parameters-320546
